### PR TITLE
Site Editor: Fix Classic Editor Block

### PR DESCRIPTION
**Pull request by david-szabo97**
## Description
Classic editor is right now broken in `master` due to missing `tinymce` scripts.
image

This is easily fixed by enqueuing the `tinymce` scripts included with WordPress.

## How has this been tested?
- Open site editor
- Add Classic block
- Make sure it works

## Screenshots
**master**
image

**this branch**
image

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by ellatrix** [comment]
Do we want to enable TinyMCE and load all its scripts in the site editor? I feel like the **legacy** editor, which has only been available for posts until now, should not be available in new contexts. Maybe for "freeform", we can enable a preview block with `setFreeformContentHandlerName`.

**Comment by ellatrix** [comment]
Actually, thinking more about it, we shouldn't allow inserting freeform HTML at all in templates. It should purely be block.

**Comment by david-szabo97** [comment]
Probably we should disable Classic block in the Site Editor in that case.

**Comment by ellatrix** [comment]
Yes, let's disable it instead :)

**Comment by gziolo** [comment]
100% agree that the Classic block shouldn't be a thing outside of the edit post page.

**Comment by david-szabo97** [comment]
I'm not quite sure if this is the best way to get rid of Classic block in full site editing. Looking forward to reviews.

`window.wp.oldEditor` is used in multiple instances to check if `Classic` block should be registered / used. So I thought getting rid of that property is the best way to make sure we totally disable `Classic` block.

**Comment by ellatrix** [comment]
Sound good! We shouldn't load any old editor scripts.

**Review by Copons — APPROVED** [review]
This works as advertised!

I'm not very fond of `delete` but I guess in this particular case beats all the alternatives.

**Review by noahtallen — COMMENTED** [review]
Rather than deleting the key here, is there a way we can prevent the script from being enqueued in the first place via PHP? I see that core WordPress adds `oldEditor` here:

https://github.com/WordPress/wordpress-develop/blob/d519284b5c4b4fa5d073498c0d88b77a4b6b8540/src/wp-includes/script-loader.php#L377-L381

Though, it doesn't seem very straightforward to delete it since it's an inline script. That means we'd have to add another inline script to "undo" what that does, and that doesn't seem like much of an improvement.

Though, I do wonder if we need to be enqueueing the "editor" script. not sure why that is being set to `oldEditor`.

**Comment by david-szabo97** [comment]
> Though, I do wonder if we need to be enqueueing the "editor" script. not sure why that is being set to oldEditor.

If we remove `editor` script then the editor doesn't load at all. It's still very confusing why it is called editor and why we set it to `oldEditor`. I think this solution will be fine for now then we can investigate further why this is so confusing.

**Review by ellatrix — CHANGES_REQUESTED** [review]
Can we try to avoid loading the old editor entirely?

I believed these are the lines to delete, which adds it as a dependency to the block library.

https://github.com/WordPress/wordpress-develop/blob/d519284b5c4b4fa5d073498c0d88b77a4b6b8540/src/wp-includes/script-loader.php#L254-L256

**Comment by ellatrix** [comment]
In other words, in the Gutenberg plugin we need to remove the `editor` dependency for the site editor. This should be possible with a filter.

> It's still very confusing why it is called editor and why we set it to `oldEditor`.

That's to avoid a conflict on the front-end, where the new block editor also uses the `wp.editor` namespace for one of the packages.

**Comment by david-szabo97** [comment]
Actually, I got it working and it resulted in less JS loaded. I pushed it in one commit so we can revert it if it seems to cause any problems. I did a few smoke testing and so far everything is working and classic editor is disabled.

File: lib/client-assets.php [hunk 1]
```diff
@@ -262,7 +262,7 @@ function gutenberg_register_packages_scripts( $scripts ) {
 		// Add dependencies that cannot be detected and generated by build tools.
 		switch ( $handle ) {
 			case 'wp-block-library':
-				array_push( $dependencies, 'editor' );
+				array_push( $dependencies, 'wp-dom-ready' );
```
Comment (noahtallen) [inline lib/client-assets.php:265]: Could you explain this change a bit more? I'm curious why we need to remove editor but add something else!

**Review by noahtallen — COMMENTED** [review]
This looks like a great approach to me. I am also unsure what side effects it could have.

File: lib/client-assets.php (hunk 1 above)
Comment (david-szabo97) [reply lib/client-assets.php:265]: The old `editor` script no longer exists so we can't have that as a dependency. If we don't change this to `wp-dom-ready` then Gutenberg doesn't load at all.

**Review by noahtallen — APPROVED** [review]
Let's try it!

One suggestion would be to also extend this behavior to the post editor when editing posts with CPT wp_template or wp_template_part. Otherwise people might insert it there and wonder why it doesn't work in site editor

**Comment by carlomanf** [comment]
The site editor edits posts too, doesn't it? What happens if a post contains classic content and is opened in the site editor?

**Comment by noahtallen** [comment]
> What happens if a post contains classic content and is opened in the site editor

Ooh, good point. 🤦 Yeah, that totally makes the approach here unviable.

**Comment by david-szabo97** [comment]
> The site editor edits posts too, doesn't it? What happens if a post contains classic content and is opened in the site editor?

Ooof. This means we need to leave the classic block in the site editor if we want to let users edit classic blocks inside the `Post Content` block. Just tried this on `master` and receiving the same error for classic blocks inside the `Post Content` block.

image

**Comment by ellatrix** [comment]
@david-szabo97 could we just preview the content (load the HTML) in that case?

* In the post editor, freeform content can be edited with the classic block.
* In the site editor, when editing a post, freeform content is just previewed, maybe can be converted to blocks.

To do that I think we'll have to register a new block and set `setFreeformContentHandlerName` with it.
This could even be useful for the block editor in general, so environments where TinyMCE isn't loaded but freeform content is present are accounted for.

**Comment by ellatrix** [comment]
Created an alternative to this PR: #27059.

**Comment by david-szabo97** [comment]
@ellatrix Why is your alternative better than truly removing the rich editor scripts? Or just simply deleting the `oldEditor` key?

This PR and yours right now do the same - disabling Classic block. (My PR desc isn't updated)

**Comment by noahtallen** [comment]
> @david-szabo97 could we just preview the content (load the HTML) in that case?

I think there is still a use case where are editing the post content in the same editor as FSE. I know this isn't our main flow right now (which focuses on templates), but it seems like an interaction that would be possible in the future.

File: lib/edit-site-page.php [hunk 2]
```diff
@@ -198,9 +198,24 @@ function gutenberg_edit_site_init( $hook ) {
...
+	wp_deregister_script( 'mce-view' );
 }
 add_action( 'admin_enqueue_scripts', 'gutenberg_edit_site_init' );
 
+/**
+ * Disallow rich editing in the site editor.
+ * This prevents tinymce from loading.
+ */
+function gutenberg_disable_richedit_tinymce() {
+	return ! gutenberg_is_edit_site_page( get_current_screen()->id );
```
Comment (ellatrix) [inline lib/edit-site-page.php:215]: What's the difference with the above? What will this do?

**Comment by ellatrix** [comment]
What happens when you load a post in the site editor containing freeform content (legacy content)?

**Comment by gziolo** [comment]
> Why is your alternative better than truly removing the rich editor scripts? Or just simply deleting the `oldEditor` key?

Ideally, we decouple Classic block from the core blocks library and enqueue it separately on the edit post page. It would have to be explored as it’s hard to tell what implications it creates. As @ellatrix pointed out, scripts deregistration will break all 3rd party plugins that depend on them and integrate (or plan to) with the edit site page.

File: lib/edit-site-page.php (hunk 2 above)
Comment (david-szabo97) [reply lib/edit-site-page.php:215]: Some TinyMCE inline scripts can't be removed by deregistering the scripts. But they are tied to this filter so by returning false here we can prevent them from loading.

File: lib/edit-site-page.php [hunk 3]
```diff
@@ -198,9 +198,24 @@ function gutenberg_edit_site_init( $hook ) {
 	wp_enqueue_script( 'wp-format-library' );
 	wp_enqueue_style( 'wp-edit-site' );
 	wp_enqueue_style( 'wp-format-library' );
+
+	wp_deregister_script( 'editor' );
+	wp_deregister_script( 'wp-tinymce' );
+	wp_deregister_script( 'wp-tinymce-root' );
+	wp_deregister_script( 'wp-tinymce-lists' );
+	wp_deregister_script( 'mce-view' );
```
[the remark this answers is not in the document]
Comment (david-szabo97) [reply lib/edit-site-page.php:206]: Good point

**Comment by david-szabo97** [comment]
Superseded by https://github.com/WordPress/gutenberg/pull/27059

**Pull request closed without merging**